demos/fleet_provisioning/fleet_provisioning_with_csr/fleet_provisioning_serializer.h

```diff
@@ -0,0 +1,112 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/**
+ * This file declares functions for serializing and parsing CBOR encoded Fleet
+ * Provisioning API payloads.
+ */
+
+/* Standard includes. */
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+/**
+ * @brief Creates the request payload to be published to the
+ * CreateCertificateFromCsr API in order to request a certificate from AWS IoT
+ * for the included Certificate Signing Request (CSR).
+ *
+ * @param[in] pBuffer Buffer into which to write the publish request payload.
+ * @param[in] bufferLength Length of #pBuffer.
+ * @param[in] pCsr The CSR to include in the request payload.
+ * @param[in] csrLength The length of #pCsr.
+ * @param[out] pOutLengthWritten The length of the publish request payload.
+ */
+bool generateCsrRequest( uint8_t * pBuffer,
+                         size_t bufferLength,
+                         const char * pCsr,
+                         size_t csrLength,
+                         size_t * pOutLengthWritten );
+
+/**
+ * @brief Creates the request payload to be published to the RegisterThing API
+ * in order to activate the provisioned certificate and receive a Thing name.
+ *
+ * @param[in] pBuffer Buffer into which to write the publish request payload.
+ * @param[in] bufferLength Length of #buffer.
+ * @param[in] pCertificateOwnershipToken The certificate's certificate
+ * ownership token.
+ * @param[in] certificateOwnershipTokenLength Length of
+ * #certificateOwnershipToken.
+ * @param[out] pOutLengthWritten The length of the publish request payload.
+ */
+bool generateRegisterThingRequest( uint8_t * pBuffer,
+                                   size_t bufferLength,
+                                   const char * pCertificateOwnershipToken,
+                                   size_t certificateOwnershipTokenLength,
+                                   const char * pSerial,
+                                   size_t serialLength,
+                                   size_t * pOutLengthWritten );
+
+/**
+ * @brief Extracts the certificate, certificate ID, and certificate ownership
+ * token from a CreateCertificateFromCsr accepted response. These are copied
+ * the provided buffers so that they can outlive the data in the response
+ * buffer and as CBOR strings may be chunked.
```

Comment:
```suggestion
 * token from a CreateCertificateFromCsr accepted response. These are copied
 * to the provided buffers so that they can outlive the data in the response
 * buffer and as CBOR strings may be chunked.
```